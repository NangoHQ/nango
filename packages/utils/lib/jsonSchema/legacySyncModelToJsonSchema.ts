import type { LegacySyncModelSchema } from '@nangohq/types';
import type { JSONSchema7 } from 'json-schema';

/**
 * Converts a list of LegacySyncModelSchema to a JSON Schema with all the schemas stored in the definitions property.
 * @param models Array of LegacySyncModelSchema
 */
export function legacySyncModelsToJsonSchema(models: LegacySyncModelSchema[]): JSONSchema7 {
    const definitions: Record<string, JSONSchema7> = {};
    for (const model of models) {
        definitions[model.name] = legacySyncModelToJsonSchema(model, models);
    }
    return { definitions };
}

function legacySyncModelToJsonSchema(model: LegacySyncModelSchema, allModels: LegacySyncModelSchema[]): JSONSchema7 {
    const properties = new Map<string, JSONSchema7>();
    const required = new Set<string>();

    const fields = Array.isArray(model.fields) ? model.fields : [];

    for (const field of fields) {
        // Optional fields end with '?' or '| undefined'
        const isOptional = /\?$/.test(field.name) || (typeof field.type === 'string' && /\|\s*undefined$/.test(field.type));
        // Remove the optional marker.
        const cleanName = field.name.replace(/\?$/, '');

        const nameChunks = cleanName.split('.');
        // Handling nested fields represented by a dot in the name.
        if (nameChunks.length > 1) {
            const parentName = nameChunks.shift()!;
            const rest = nameChunks.join('.');

            // We call this function recusively, but without the parent name. Then we merge resulting object schemas.
            const schema = legacySyncModelToJsonSchema({ name: '', fields: [{ ...field, name: rest }] }, allModels);

            if (properties.has(parentName)) {
                properties.set(parentName, mergeObjects(properties.get(parentName)!, schema));
            } else {
                properties.set(parentName, schema);
            }

            if (!isOptional) {
                required.add(parentName);
            }

            continue;
        }

        properties.set(cleanName, legacySyncFieldToJsonSchema(field, allModels));

        if (!isOptional) {
            required.add(cleanName);
        }
    }

    return {
        type: 'object',
        properties: Object.fromEntries(properties),
        required: Array.from(required)
    };
}

function legacySyncFieldToJsonSchema(field: { name: string; type: string }, allModels: LegacySyncModelSchema[]): JSONSchema7 {
    // While our typescript definition defines this as a string, we have many legacy models in the database
    // where this is an object or array describing submodels.
    if (typeof field.type !== 'string') {
        if (Array.isArray(field.type)) {
            if ((field.type as unknown[]).length > 1) {
                return {
                    type: 'array',
                    items: {
                        oneOf: (field.type as unknown[]).map((t) =>
                            legacySyncModelToJsonSchema({ name: '', fields: objectToFields(t as Record<string, string>) }, allModels)
                        )
                    }
                };
            }
            if ((field.type as unknown[]).length === 0) {
                return {
                    type: 'array',
                    items: {}
                };
            }
            return {
                type: 'array',
                items: legacySyncModelToJsonSchema({ name: '', fields: objectToFields(field.type[0] as Record<string, string>) }, allModels)
            };
        }

        return legacySyncModelToJsonSchema({ name: '', fields: objectToFields(field.type as Record<string, string>) }, allModels);
    }

    let typeStr = field.type?.trim() ?? '';
    // Handle optional
    typeStr = typeStr.replace(/\|\s*undefined$/, '').trim();
    // Handle union
    if (typeStr.includes('|')) {
        const types = typeStr
            .split('|')
            .map((t) => t.trim())
            .filter((t) => t !== 'undefined');
        return {
            oneOf: types.map((t) => legacySyncTypeToJsonSchema(t, allModels))
        };
    }
    return legacySyncTypeToJsonSchema(typeStr, allModels);
}

function legacySyncTypeToJsonSchema(typeStr: string, allModels: LegacySyncModelSchema[]): JSONSchema7 {
    // Array type
    const arrayMatch = typeStr.match(/^(.*)\[\]$/);
    if (arrayMatch && typeof arrayMatch[1] === 'string') {
        const itemType = arrayMatch[1].trim();
        return {
            type: 'array',
            items: legacySyncTypeToJsonSchema(itemType, allModels)
        };
    }

    // Primitive types
    if (legacyPrimitiveTypeMap[typeStr]) {
        return legacyPrimitiveTypeMap[typeStr];
    }

    if (allModels.find((m) => m.name === typeStr)) {
        return { $ref: `#/definitions/${typeStr}` };
    }

    return { type: 'string', const: typeStr };
}

const legacyPrimitiveTypeMap: Record<string, JSONSchema7> = {
    integer: { type: 'integer' },
    int: { type: 'integer' },
    float: { type: 'number' },
    number: { type: 'number' },
    boolean: { type: 'boolean' },
    bool: { type: 'boolean' },
    true: { type: 'boolean', const: true },
    false: { type: 'boolean', const: false },
    string: { type: 'string' },
    char: { type: 'string' },
    varchar: { type: 'string' },
    date: { type: 'string', format: 'date-time' },
    null: { type: 'null' },
    undefined: { type: 'null' },
    any: {},
    object: { type: 'object' },
    array: { type: 'array', items: {} }
};

function objectToFields(object: Record<string, string>): LegacySyncModelSchema['fields'] {
    const fields: LegacySyncModelSchema['fields'] = [];
    for (const [key, value] of Object.entries(object)) {
        fields.push({ name: key, type: value });
    }
    return fields;
}

function mergeObjects(a: JSONSchema7, b: JSONSchema7): JSONSchema7 {
    // If this happens it's invalid anyway.
    if (a.type !== 'object' || b.type !== 'object') {
        return a;
    }

    return {
        ...a,
        properties: { ...a.properties, ...b.properties },
        required: [...(a.required || []), ...(b.required || [])]
    };
}
