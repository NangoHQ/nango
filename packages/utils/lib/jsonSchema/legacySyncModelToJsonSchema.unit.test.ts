import { describe, expect, it } from 'vitest';

import { legacySyncModelsToJsonSchema } from './legacySyncModelToJsonSchema.js';

import type { LegacySyncModelSchema } from '@nangohq/types';

describe('legacySyncModelsToJsonSchema', () => {
    it('should handle all primitive types and basic features', () => {
        const models: LegacySyncModelSchema[] = [
            {
                name: 'AllPrimitiveTypes',
                fields: [
                    { name: 'stringField', type: 'string' },
                    { name: 'numberField', type: 'number' },
                    { name: 'booleanField', type: 'boolean' },
                    { name: 'dateField', type: 'date' },
                    { name: 'optionalString', type: 'string | undefined' },
                    { name: 'unknownType', type: 'uuid' }, // Should default to string
                    { name: 'stringArray', type: 'string[]' },
                    { name: 'numberArray', type: 'number[] | undefined' }
                ]
            }
        ];
        const result = legacySyncModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle references to other models', () => {
        const models: LegacySyncModelSchema[] = [
            {
                name: 'User',
                fields: [
                    { name: 'id', type: 'string' },
                    { name: 'profile', type: 'Profile' },
                    { name: 'roles', type: 'Role[]' }
                ]
            },
            {
                name: 'Profile',
                fields: [
                    { name: 'bio', type: 'string' },
                    { name: 'avatar', type: 'string | null' }
                ]
            },
            {
                name: 'Role',
                fields: [{ name: 'name', type: 'string' }]
            }
        ];
        const result = legacySyncModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle unions and optionals', () => {
        const models: LegacySyncModelSchema[] = [
            {
                name: 'Event',
                fields: [
                    { name: 'id', type: 'string' },
                    { name: 'status', type: 'active | canceled' },
                    { name: 'maybeString', type: 'string | null | undefined' },
                    { name: 'maybeModel', type: 'Profile | null' }
                ]
            },
            {
                name: 'Profile',
                fields: [{ name: 'bio', type: 'string' }]
            }
        ];
        const result = legacySyncModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle arrays of models and primitives', () => {
        const models: LegacySyncModelSchema[] = [
            {
                name: 'Document',
                fields: [
                    { name: 'id', type: 'string' },
                    { name: 'tags', type: 'string[]' },
                    { name: 'collaborators', type: 'User[] | undefined' }
                ]
            },
            {
                name: 'User',
                fields: [
                    { name: 'id', type: 'string' },
                    { name: 'email', type: 'string' }
                ]
            }
        ];
        const result = legacySyncModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle empty models array', () => {
        const models: LegacySyncModelSchema[] = [];
        const result = legacySyncModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle model with no fields', () => {
        const models: LegacySyncModelSchema[] = [
            {
                name: 'EmptyModel',
                fields: []
            }
        ];
        const result = legacySyncModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should still reference a non-existent model', () => {
        const models: LegacySyncModelSchema[] = [
            {
                name: 'HasMissingRef',
                fields: [
                    { name: 'id', type: 'string' },
                    { name: 'missing', type: 'NonExistentModel' }
                ]
            }
        ];
        const result = legacySyncModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });
});
