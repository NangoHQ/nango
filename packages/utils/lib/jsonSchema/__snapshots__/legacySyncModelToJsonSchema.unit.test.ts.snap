// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`legacySyncModelsToJsonSchema > real-world examples > GoogleCalendarEvent 1`] = `
{
  "definitions": {
    "GoogleCalendarEvent": {
      "properties": {
        "attachments": {
          "properties": {
            "0": {
              "properties": {
                "fileId": {
                  "type": "string",
                },
                "fileUrl": {
                  "type": "string",
                },
                "iconLink": {
                  "type": "string",
                },
                "mimeType": {
                  "type": "string",
                },
                "title": {
                  "type": "string",
                },
              },
              "required": [
                "title",
                "fileId",
                "fileUrl",
                "iconLink",
                "mimeType",
              ],
              "type": "object",
            },
          },
          "required": [
            "0",
          ],
          "type": "object",
        },
        "attendees": {
          "properties": {
            "0": {
              "properties": {
                "additionalGuests": {
                  "type": "integer",
                },
                "comment": {
                  "type": "string",
                },
                "displayName": {
                  "type": "string",
                },
                "email": {
                  "type": "string",
                },
                "id": {
                  "type": "string",
                },
                "optional": {
                  "type": "boolean",
                },
                "organizer": {
                  "type": "boolean",
                },
                "resource": {
                  "type": "boolean",
                },
                "responseStatus": {
                  "type": "string",
                },
                "self": {
                  "type": "boolean",
                },
              },
              "required": [
                "id",
                "self",
                "email",
                "comment",
                "optional",
                "resource",
                "organizer",
                "displayName",
                "responseStatus",
                "additionalGuests",
              ],
              "type": "object",
            },
          },
          "required": [
            "0",
          ],
          "type": "object",
        },
        "conferenceData": {
          "properties": {
            "conferenceSolution": {
              "properties": {
                "iconUri": {
                  "type": "string",
                },
                "key": {
                  "properties": {
                    "type": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "type",
                  ],
                  "type": "object",
                },
                "name": {
                  "type": "string",
                },
              },
              "required": [
                "key",
                "name",
                "iconUri",
              ],
              "type": "object",
            },
            "createRequest": {
              "properties": {
                "conferenceSolutionKey": {
                  "properties": {
                    "type": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "type",
                  ],
                  "type": "object",
                },
                "requestId": {
                  "type": "string",
                },
                "status": {
                  "properties": {
                    "statusCode": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "statusCode",
                  ],
                  "type": "object",
                },
              },
              "required": [
                "status",
                "requestId",
                "conferenceSolutionKey",
              ],
              "type": "object",
            },
            "entryPoints": {
              "items": {
                "properties": {
                  "accessCode": {
                    "type": "string",
                  },
                  "entryPointType": {
                    "type": "string",
                  },
                  "label": {
                    "type": "string",
                  },
                  "meetingCode": {
                    "type": "string",
                  },
                  "passcode": {
                    "type": "string",
                  },
                  "password": {
                    "type": "string",
                  },
                  "pin": {
                    "type": "string",
                  },
                  "uri": {
                    "type": "string",
                  },
                },
                "required": [
                  "pin",
                  "uri",
                  "label",
                  "passcode",
                  "password",
                  "accessCode",
                  "meetingCode",
                  "entryPointType",
                ],
                "type": "object",
              },
              "type": "array",
            },
          },
          "required": [
            "createRequest",
            "entryPoints",
            "conferenceSolution",
          ],
          "type": "object",
        },
        "creator": {
          "properties": {
            "displayName": {
              "type": "string",
            },
            "email": {
              "type": "string",
            },
            "id": {
              "type": "string",
            },
            "self": {
              "type": "boolean",
            },
          },
          "required": [
            "id",
            "email",
            "displayName",
            "self",
          ],
          "type": "object",
        },
        "end": {
          "properties": {
            "date": {
              "format": "date-time",
              "type": "string",
            },
            "string": {
              "type": "string",
            },
            "timeZone": {
              "type": "string",
            },
          },
          "required": [
            "date",
            "string",
            "timeZone",
          ],
          "type": "object",
        },
        "endTimeUnspecified": {
          "type": "boolean",
        },
        "eventType": {
          "type": "string",
        },
        "extendedProperties": {
          "properties": {
            "private": {
              "properties": {
                "key": {
                  "type": "string",
                },
              },
              "required": [
                "key",
              ],
              "type": "object",
            },
            "shared": {
              "properties": {
                "key": {
                  "type": "string",
                },
              },
              "required": [
                "key",
              ],
              "type": "object",
            },
          },
          "required": [
            "private",
            "shared",
          ],
          "type": "object",
        },
        "gadget": {
          "properties": {
            "preferences": {
              "properties": {
                "key": {
                  "type": "string",
                },
              },
              "required": [
                "key",
              ],
              "type": "object",
            },
          },
          "required": [
            "preferences",
          ],
          "type": "object",
        },
        "organizer": {
          "properties": {
            "displayName": {
              "type": "string",
            },
            "email": {
              "type": "string",
            },
            "id": {
              "type": "string",
            },
            "self": {
              "type": "boolean",
            },
          },
          "required": [
            "id",
            "email",
            "displayName",
            "self",
          ],
          "type": "object",
        },
        "originalStartTime": {
          "properties": {
            "date": {
              "format": "date-time",
              "type": "string",
            },
            "dateTime": {
              "type": "string",
            },
            "timeZone": {
              "type": "string",
            },
          },
          "required": [
            "date",
            "dateTime",
            "timeZone",
          ],
          "type": "object",
        },
        "recurrence": {
          "properties": {
            "0": {
              "type": "string",
            },
          },
          "required": [
            "0",
          ],
          "type": "object",
        },
        "recurringEventId": {
          "type": "string",
        },
        "reminders": {
          "properties": {
            "overrides": {
              "items": {
                "properties": {
                  "method": {
                    "type": "string",
                  },
                  "minutes": {
                    "type": "integer",
                  },
                },
                "required": [
                  "method",
                  "minutes",
                ],
                "type": "object",
              },
              "type": "array",
            },
          },
          "required": [
            "overrides",
          ],
          "type": "object",
        },
        "source": {
          "properties": {
            "title": {
              "type": "string",
            },
            "url": {
              "type": "string",
            },
          },
          "required": [
            "url",
            "title",
          ],
          "type": "object",
        },
        "start": {
          "properties": {
            "date": {
              "format": "date-time",
              "type": "string",
            },
            "dateTime": {
              "type": "string",
            },
            "timeZone": {
              "type": "string",
            },
          },
          "required": [
            "date",
            "dateTime",
            "timeZone",
          ],
          "type": "object",
        },
        "workingLocationProperties": {
          "properties": {
            "customLocation": {
              "properties": {
                "label": {
                  "type": "string",
                },
              },
              "required": [
                "label",
              ],
              "type": "object",
            },
            "homeOffice": {
              "type": "string",
            },
            "officeLocation": {
              "properties": {
                "buildingId": {
                  "type": "string",
                },
                "deskId": {
                  "type": "string",
                },
                "floorId": {
                  "type": "string",
                },
                "floorSectionId": {
                  "type": "string",
                },
                "label": {
                  "type": "string",
                },
              },
              "required": [
                "label",
                "deskId",
                "floorId",
                "buildingId",
                "floorSectionId",
              ],
              "type": "object",
            },
            "type": {
              "type": "string",
            },
          },
          "required": [
            "type",
            "homeOffice",
            "customLocation",
            "officeLocation",
          ],
          "type": "object",
        },
      },
      "required": [
        "creator",
        "organizer",
        "start",
        "end",
        "endTimeUnspecified",
        "recurrence",
        "recurringEventId",
        "originalStartTime",
        "attendees",
        "extendedProperties",
        "conferenceData",
        "gadget",
        "reminders",
        "source",
        "workingLocationProperties",
        "attachments",
        "eventType",
      ],
      "type": "object",
    },
  },
}
`;

exports[`legacySyncModelsToJsonSchema > should handle all primitive types and basic features 1`] = `
{
  "definitions": {
    "AllPrimitiveTypes": {
      "properties": {
        "anyField": {},
        "arrayField": {
          "items": {},
          "type": "array",
        },
        "boolField": {
          "type": "boolean",
        },
        "booleanField": {
          "type": "boolean",
        },
        "charField": {
          "type": "string",
        },
        "dateField": {
          "format": "date-time",
          "type": "string",
        },
        "falseField": {
          "const": false,
          "type": "boolean",
        },
        "floatField": {
          "type": "number",
        },
        "intField": {
          "type": "integer",
        },
        "integerField": {
          "type": "integer",
        },
        "literalUnion": {
          "oneOf": [
            {
              "const": "red",
              "type": "string",
            },
            {
              "const": "blue",
              "type": "string",
            },
          ],
        },
        "nullField": {
          "type": "null",
        },
        "numberArray": {
          "items": {
            "type": "number",
          },
          "type": "array",
        },
        "numberField": {
          "type": "number",
        },
        "objectField": {
          "type": "object",
        },
        "optionalString": {
          "type": "string",
        },
        "stringArray": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
        "stringField": {
          "type": "string",
        },
        "trueField": {
          "const": true,
          "type": "boolean",
        },
        "undefinedField": {
          "type": "null",
        },
        "unknownType": {
          "const": "uuid",
          "type": "string",
        },
        "varcharField": {
          "type": "string",
        },
      },
      "required": [
        "stringField",
        "charField",
        "varcharField",
        "numberField",
        "floatField",
        "intField",
        "integerField",
        "booleanField",
        "boolField",
        "trueField",
        "falseField",
        "dateField",
        "undefinedField",
        "nullField",
        "anyField",
        "objectField",
        "arrayField",
        "unknownType",
        "stringArray",
        "literalUnion",
      ],
      "type": "object",
    },
  },
}
`;

exports[`legacySyncModelsToJsonSchema > should handle arrays of models and primitives 1`] = `
{
  "definitions": {
    "Document": {
      "properties": {
        "collaborators": {
          "items": {
            "$ref": "#/definitions/User",
          },
          "type": "array",
        },
        "id": {
          "type": "string",
        },
        "tags": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
      },
      "required": [
        "id",
        "tags",
      ],
      "type": "object",
    },
    "User": {
      "properties": {
        "email": {
          "type": "string",
        },
        "id": {
          "type": "string",
        },
      },
      "required": [
        "id",
        "email",
      ],
      "type": "object",
    },
  },
}
`;

exports[`legacySyncModelsToJsonSchema > should handle empty models array 1`] = `
{
  "definitions": {},
}
`;

exports[`legacySyncModelsToJsonSchema > should handle inline objects 1`] = `
{
  "definitions": {
    "InlineObject": {
      "properties": {
        "object": {
          "properties": {
            "number": {
              "type": "number",
            },
            "string": {
              "type": "string",
            },
          },
          "required": [
            "string",
            "number",
          ],
          "type": "object",
        },
        "string": {
          "type": "string",
        },
      },
      "required": [
        "string",
        "object",
      ],
      "type": "object",
    },
  },
}
`;

exports[`legacySyncModelsToJsonSchema > should handle model with no fields 1`] = `
{
  "definitions": {
    "EmptyModel": {
      "properties": {},
      "required": [],
      "type": "object",
    },
  },
}
`;

exports[`legacySyncModelsToJsonSchema > should handle nested fields separated by dots 1`] = `
{
  "definitions": {
    "NestedFields": {
      "properties": {
        "nested": {
          "properties": {
            "number": {
              "type": "number",
            },
            "object": {
              "properties": {
                "inlineObject": {
                  "properties": {
                    "number": {
                      "type": "number",
                    },
                    "string": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "string",
                    "number",
                  ],
                  "type": "object",
                },
              },
              "required": [
                "inlineObject",
              ],
              "type": "object",
            },
            "string": {
              "type": "string",
            },
          },
          "required": [
            "string",
            "number",
            "object",
            "object",
            "object",
          ],
          "type": "object",
        },
        "string": {
          "type": "string",
        },
      },
      "required": [
        "string",
        "nested",
      ],
      "type": "object",
    },
  },
}
`;

exports[`legacySyncModelsToJsonSchema > should handle optionals from name 1`] = `
{
  "definitions": {
    "Model": {
      "properties": {
        "optionalNumber": {
          "type": "number",
        },
        "optionalString": {
          "type": "string",
        },
        "requiredNumber": {
          "type": "number",
        },
        "requiredString": {
          "type": "string",
        },
      },
      "required": [
        "requiredString",
        "requiredNumber",
      ],
      "type": "object",
    },
  },
}
`;

exports[`legacySyncModelsToJsonSchema > should handle references to other models 1`] = `
{
  "definitions": {
    "Profile": {
      "properties": {
        "avatar": {
          "oneOf": [
            {
              "type": "string",
            },
            {
              "type": "null",
            },
          ],
        },
        "bio": {
          "type": "string",
        },
      },
      "required": [
        "bio",
        "avatar",
      ],
      "type": "object",
    },
    "Role": {
      "properties": {
        "name": {
          "type": "string",
        },
      },
      "required": [
        "name",
      ],
      "type": "object",
    },
    "User": {
      "properties": {
        "id": {
          "type": "string",
        },
        "profile": {
          "$ref": "#/definitions/Profile",
        },
        "roles": {
          "items": {
            "$ref": "#/definitions/Role",
          },
          "type": "array",
        },
      },
      "required": [
        "id",
        "profile",
        "roles",
      ],
      "type": "object",
    },
  },
}
`;

exports[`legacySyncModelsToJsonSchema > should handle unions and optionals 1`] = `
{
  "definitions": {
    "Event": {
      "properties": {
        "id": {
          "type": "string",
        },
        "maybeModel": {
          "oneOf": [
            {
              "$ref": "#/definitions/Profile",
            },
            {
              "type": "null",
            },
          ],
        },
        "maybeString": {
          "oneOf": [
            {
              "type": "string",
            },
            {
              "type": "null",
            },
          ],
        },
        "status": {
          "oneOf": [
            {
              "const": "active",
              "type": "string",
            },
            {
              "const": "canceled",
              "type": "string",
            },
          ],
        },
      },
      "required": [
        "id",
        "status",
        "maybeModel",
      ],
      "type": "object",
    },
    "Profile": {
      "properties": {
        "bio": {
          "type": "string",
        },
      },
      "required": [
        "bio",
      ],
      "type": "object",
    },
  },
}
`;

exports[`legacySyncModelsToJsonSchema > should not reference a non-existent model 1`] = `
{
  "definitions": {
    "HasMissingRef": {
      "properties": {
        "id": {
          "type": "string",
        },
        "missing": {
          "const": "NonExistentModel",
          "type": "string",
        },
      },
      "required": [
        "id",
        "missing",
      ],
      "type": "object",
    },
  },
}
`;
